File: apps/frontend/src/lib/api/viewer.ts
import { createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import { ViewerService } from '$lib/pb/chatto/api/v1/viewer_connect';
import { PresenceStatus as APIPresenceStatus } from '$lib/pb/chatto/api/v1/presence_pb';
import { NotificationLevel as APINotificationLevel } from '$lib/pb/chatto/api/v1/notification_preferences_pb';
import { TimeFormat as APITimeFormat } from '$lib/pb/chatto/api/v1/viewer_pb';
import { NotificationLevel, PresenceStatus, TimeFormat } from '$lib/render/types';

export type ViewerAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
};

export type CurrentUser = {
  id: string;
  login: string;
  displayName: string;
  avatarUrl?: string | null;
  customStatus?: {
    emoji: string;
    text: string;
    expiresAt?: string | null;
  } | null;
  presenceStatus: PresenceStatus;
  hasVerifiedEmail: boolean;
  viewerCanDeleteAccount: boolean;
  lastLoginChange?: string | null;
  settings?: {
    timezone?: string | null;
    timeFormat: TimeFormat;
  } | null;
};

export type ViewerCapabilities = {
  canViewAdmin: boolean;
  canStartDMs: boolean;
  canAdminViewUsers: boolean;
  canAdminManageUsers: boolean;
  canAdminViewRoles: boolean;
  canAdminManageRoles: boolean;
  canAdminViewSystem: boolean;
  canAdminViewAudit: boolean;
  canManageUserPermissions: boolean;
};

export type NotificationPreference = {
  level: NotificationLevel;
  effectiveLevel: NotificationLevel;
};

export type RoomNotificationPreference = NotificationPreference & {
  roomId: string;
};

export type ViewerState = ViewerCapabilities & {
  user: CurrentUser;
  serverNotificationPreference: NotificationPreference;
  roomNotificationPreferences: RoomNotificationPreference[];
};

export async function getViewerStateViaConnect(config: ViewerAPIConfig): Promise<ViewerState> {
  const transport = createConnectTransport({
    baseUrl: config.baseUrl,
    useBinaryFormat: true
  });
  const client = createClient(ViewerService, transport);
  const response = await client.getViewer(
    {},
    {
      headers: config.bearerToken ? { Authorization: `Bearer ${config.bearerToken}` } : undefined
    }
  );
  if (!response.user) {
    throw new Error('viewer response did not include a user');
  }
  if (!response.user.profile) {
    throw new Error('viewer response did not include a user profile');
  }
  const user = response.user.profile;
  if (!user.user) {
    throw new Error('viewer response did not include a user summary');
  }
  const summary = user.user;
  const capabilities = response.capabilities;
  return {
    user: {
      id: summary.id,
      login: summary.login,
      displayName: summary.displayName,
      avatarUrl: summary.avatarUrl ?? null,
      customStatus: user.customStatus
        ? {
            emoji: user.customStatus.emoji,
            text: user.customStatus.text,
            expiresAt: user.customStatus.expiresAt?.toDate().toISOString() ?? null
          }
        : null,
      presenceStatus: apiPresenceStatus(user.presenceStatus),
      hasVerifiedEmail: response.user.hasVerifiedEmail,
      viewerCanDeleteAccount: response.user.viewerCanDeleteAccount ?? false,
      lastLoginChange: response.user.lastLoginChange?.toDate().toISOString() ?? null,
      settings: response.user.settings
        ? {
            timezone: response.user.settings.timezone ?? null,
            timeFormat: apiTimeFormat(response.user.settings.timeFormat)
          }
        : null
    },
    canViewAdmin: capabilities?.canViewAdmin ?? false,
    canStartDMs: capabilities?.canStartDms ?? false,
    canAdminViewUsers: capabilities?.canAdminViewUsers ?? false,
    canAdminManageUsers: capabilities?.canAdminManageUsers ?? false,
    canAdminViewRoles: capabilities?.canAdminViewRoles ?? false,
    canAdminManageRoles: capabilities?.canAdminManageRoles ?? false,
    canAdminViewSystem: capabilities?.canAdminViewSystem ?? false,
    canAdminViewAudit: capabilities?.canAdminViewAudit ?? false,
    canManageUserPermissions: capabilities?.canManageUserPermissions ?? false,
    serverNotificationPreference: {
      level: apiNotificationLevel(response.serverNotificationPreference?.level),
      effectiveLevel: apiNotificationLevel(response.serverNotificationPreference?.effectiveLevel)
    },
    roomNotificationPreferences: response.roomNotificationPreferences.map((pref) => ({
      roomId: pref.roomId,
      level: apiNotificationLevel(pref.level),
      effectiveLevel: apiNotificationLevel(pref.effectiveLevel)
    }))
  };
}

export async function getCurrentUserViaConnect(config: ViewerAPIConfig): Promise<CurrentUser> {
  return (await getViewerStateViaConnect(config)).user;
}

function apiNotificationLevel(level: APINotificationLevel | undefined): NotificationLevel {
  switch (level) {
    case APINotificationLevel.MUTED:
      return NotificationLevel.Muted;
    case APINotificationLevel.NORMAL:
      return NotificationLevel.Normal;
    case APINotificationLevel.ALL_MESSAGES:
      return NotificationLevel.AllMessages;
    case APINotificationLevel.DEFAULT:
    case APINotificationLevel.UNSPECIFIED:
    default:
      return NotificationLevel.Default;
  }
}

function apiPresenceStatus(status: APIPresenceStatus): PresenceStatus {
  switch (status) {
    case APIPresenceStatus.AWAY:
      return PresenceStatus.Away;
    case APIPresenceStatus.DO_NOT_DISTURB:
      return PresenceStatus.DoNotDisturb;
    case APIPresenceStatus.ONLINE:
      return PresenceStatus.Online;
    case APIPresenceStatus.OFFLINE:
    case APIPresenceStatus.UNSPECIFIED:
    default:
      return PresenceStatus.Offline;
  }
}

function apiTimeFormat(format: APITimeFormat): TimeFormat {
  switch (format) {
    case APITimeFormat.TIME_FORMAT_12_HOUR:
      return TimeFormat.TwelveHour;
    case APITimeFormat.TIME_FORMAT_24_HOUR:
      return TimeFormat.TwentyFourHour;
    case APITimeFormat.TIME_FORMAT_AUTO:
    case APITimeFormat.TIME_FORMAT_UNSPECIFIED:
    default:
      return TimeFormat.Auto;
  }
}

File: cli/internal/connectapi/viewer.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

type viewerService struct {
	api *API
}

func (s *viewerService) GetViewer(ctx context.Context, _ *connect.Request[apiv1.GetViewerRequest]) (*connect.Response[apiv1.GetViewerResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	user, err := s.api.core.GetUser(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}

	responseUser, err := s.viewerUser(ctx, user)
	if err != nil {
		return nil, err
	}
	capabilities, err := s.viewerCapabilities(ctx, caller.UserID)
	if err != nil {
		return nil, err
	}
	serverPreference, err := s.serverNotificationPreference(ctx, caller.UserID)
	if err != nil {
		return nil, err
	}
	roomPreferences, err := s.roomNotificationPreferences(ctx, caller.UserID)
	if err != nil {
		return nil, err
	}

	return connect.NewResponse(&apiv1.GetViewerResponse{
		User:                         responseUser,
		Capabilities:                 capabilities,
		ServerNotificationPreference: serverPreference,
		RoomNotificationPreferences:  roomPreferences,
	}), nil
}

func (s *viewerService) viewerUser(ctx context.Context, user *corev1.User) (*apiv1.ViewerUser, error) {
	hasVerifiedEmail, err := s.api.core.HasVerifiedEmail(ctx, user.GetId())
	if err != nil {
		return nil, connectError(err)
	}
	settings, err := s.api.core.GetUserSettings(ctx, user.GetId())
	if err != nil {
		return nil, connectError(err)
	}
	presence, err := s.api.core.GetUserPresence(ctx, user.GetId())
	if err != nil {
		return nil, connectError(err)
	}
	canDeleteAccount, err := s.api.core.CanDeleteUser(ctx, user.GetId(), user.GetId())
	if err != nil {
		return nil, connectError(err)
	}
	lastLoginChange, err := s.api.core.GetLastLoginChange(ctx, user.GetId())
	if err != nil {
		return nil, connectError(err)
	}

	response := &apiv1.ViewerUser{
		HasVerifiedEmail:       hasVerifiedEmail,
		Settings:               coreUserSettingsToAPI(settings),
		ViewerCanDeleteAccount: canDeleteAccount,
		Profile: &apiv1.UserPresenceSummary{
			User: &apiv1.UserSummary{
				Id:          user.GetId(),
				Login:       user.GetLogin(),
				DisplayName: user.GetDisplayName(),
				Deleted:     user.GetDeleted(),
			},
			CustomStatus:   coreCustomStatusToAPI(user.GetCustomStatus()),
			PresenceStatus: corePresenceStatusToAPI(presence),
		},
	}
	if avatarURL, err := s.api.core.GetUserAvatarURL(ctx, user.GetId(), nil, nil, ""); err != nil {
		return nil, connectError(err)
	} else if avatarURL != "" {
		response.Profile.User.AvatarUrl = stringPtr(s.api.absolutizeAssetURL(ctx, avatarURL))
	}
	if !lastLoginChange.IsZero() {
		response.LastLoginChange = timestamppb.New(lastLoginChange)
	}

	return response, nil
}

func (s *viewerService) viewerCapabilities(ctx context.Context, userID string) (*apiv1.ViewerCapabilities, error) {
	canViewAdmin, err := s.api.core.HasAnyAdminPermission(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canStartDMs, err := s.api.core.CanStartDM(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canAdminViewUsers, err := s.api.core.CanAdminUsersView(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canAdminManageUsers, err := s.api.core.CanAssignRoles(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canAdminManageRoles, err := s.api.core.CanManageRoles(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canManageUserPermissions, err := s.api.core.CanManageUserPermissions(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canAdminViewSystem, err := s.api.core.CanAdminSystemView(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	canAdminViewAudit, err := s.api.core.CanAdminAuditView(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	hasUnreadFollowedThreads, err := s.api.core.HasUnreadFollowedThreads(ctx, userID, []string{core.LegacySpaceIDForRoomKind(core.KindChannel)})
	if err != nil {
		return nil, connectError(err)
	}

	return &apiv1.ViewerCapabilities{
		CanViewAdmin:             canViewAdmin,
		CanStartDms:              canStartDMs,
		CanAdminViewUsers:        canAdminViewUsers,
		CanAdminManageUsers:      canAdminManageUsers,
		CanAdminViewRoles:        canAdminManageRoles,
		CanAdminManageRoles:      canAdminManageRoles,
		CanAdminViewSystem:       canAdminViewSystem,
		CanAdminViewAudit:        canAdminViewAudit,
		HasUnreadFollowedThreads: hasUnreadFollowedThreads,
		CanManageUserPermissions: canManageUserPermissions,
	}, nil
}

func (s *viewerService) serverNotificationPreference(ctx context.Context, userID string) (*apiv1.ServerNotificationPreference, error) {
	level, err := s.api.core.GetSpaceNotificationLevel(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	effectiveLevel := level
	if effectiveLevel == corev1.NotificationLevel_NOTIFICATION_LEVEL_UNSPECIFIED {
		effectiveLevel = corev1.NotificationLevel_NOTIFICATION_LEVEL_NORMAL
	}
	return &apiv1.ServerNotificationPreference{
		Level:          coreNotificationLevelToAPI(level),
		EffectiveLevel: coreNotificationLevelToAPI(effectiveLevel),
	}, nil
}

func (s *viewerService) roomNotificationPreferences(ctx context.Context, userID string) ([]*apiv1.RoomNotificationPreference, error) {
	prefs, err := s.api.core.GetAllRoomNotificationPreferences(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	result := make([]*apiv1.RoomNotificationPreference, 0, len(prefs))
	for _, pref := range prefs {
		result = append(result, &apiv1.RoomNotificationPreference{
			RoomId:         pref.RoomID,
			Level:          coreNotificationLevelToAPI(pref.Level),
			EffectiveLevel: coreNotificationLevelToAPI(pref.EffectiveLevel),
		})
	}
	return result, nil
}

func coreUserSettingsToAPI(settings *corev1.ServerUserPreferences) *apiv1.UserSettings {
	response := &apiv1.UserSettings{TimeFormat: apiv1.TimeFormat_TIME_FORMAT_AUTO}
	if settings == nil {
		return response
	}
	if settings.Timezone != nil {
		response.Timezone = settings.Timezone
	}
	response.TimeFormat = coreTimeFormatToAPI(settings.GetTimeFormat())
	return response
}

func coreTimeFormatToAPI(format corev1.TimeFormat) apiv1.TimeFormat {
	switch format {
	case corev1.TimeFormat_TIME_FORMAT_12H:
		return apiv1.TimeFormat_TIME_FORMAT_12_HOUR
	case corev1.TimeFormat_TIME_FORMAT_24H:
		return apiv1.TimeFormat_TIME_FORMAT_24_HOUR
	default:
		return apiv1.TimeFormat_TIME_FORMAT_AUTO
	}
}

func stringPtr(value string) *string {
	return &value
}
